### ISSUE TITLE
Clippy is incompatible with serde 1.0.107

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust-clippy#8378]
### Summary

Re #7606

The version of `serde` specified in `Cargo.toml` is too low, and incompatible with this crate. Please note that `serde` doesn't follow semver rules, and `serde` version 1.0.160 is incompatible with `serde` 1.0.0.

```text
test compile_test ... FAILED

failures:

---- compile_test stdout ----
diff of stderr:

 error[E0412]: cannot find type `PhantomData` in this scope
   --> $DIR/ice-6252.rs:8:9
    |
 LL |     _n: PhantomData,
    |         ^^^^^^^^^^^ not found in this scope
    |
 help: consider importing one of these items
    |
 LL | use core::marker::PhantomData;
    |
-LL | use serde::__private::PhantomData;
+LL | use serde::export::PhantomData;
    |
 LL | use std::marker::PhantomData;
    |

 error[E0412]: cannot find type `VAL` in this scope
   --> $DIR/ice-6252.rs:10:63

```

### Reproducer

Clippy asks for serde 1.0.0, which is even lower, but currently the lowest possible version is 1.0.107, because that's what `cargo_metadata` crate specifies, and this forces this version for the whole project (Cargo doesn't duplicate dependencies with the same semver-major version).

```bash
cargo update -p serde --precise 1.0.107
cargo test --all
```

The same tests pass if you do:

```bash
cargo update -p serde
```

to bump `serde` to 1.0.136. Somewhere between 107 and 136 serde has changed its behavior.

### Version

```text
rustc 1.60.0-nightly (6abb6385b 2022-01-26)
binary: rustc
commit-hash: 6abb6385b2cb7249f67b9b3ce7522527767dd907
commit-date: 2022-01-26
host: aarch64-apple-darwin
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by flip1995** [comment]
Continuing the conversation of #7606: I mean, Clippy doesn't have a `Cargo.lock` file, so running `cargo update` if something weird breaks with dependencies is usually the solution. This probably did break for you, because you had an outdated `Cargo.lock` file in your checkout.

We could set `serde` to `1.0.125`, so that it at least uses that version. This is the version that is currently in `rust-lang/rust`s `Cargo.lock` file. Since we don't really care that much about the serde version in Clippy (as long as tests pass), I'd say we don't cause a `serde` version update in `rust-lang/rust`. We can set the minimum version to whatever works in the range of `1.0.108` and `1.0.125` though.
